File: advent/day02.py
# /usr/bin/env python3
import re

from advent.base import BaseSolver, Solution


class Solver(BaseSolver):
    def solve(self) -> Solution:
        res1 = 0
        res2 = 0
        r, g, b = 12, 13, 14
        for game in self.data.splitlines():
            possible = True
            minr, ming, minb = 0, 0, 0
            game_str, rest = game.split(":")
            game_id = int(game_str.split(" ")[1])

            for s in rest.split(";"):
                red = int((re.findall(r"(\d+) red", s) or [0])[0])
                green = int((re.findall(r"(\d+) green", s) or [0])[0])
                blue = int((re.findall(r"(\d+) blue", s) or [0])[0])

                if red > r or green > g or blue > b:
                    possible = False

                minr = max(minr, red)
                ming = max(ming, green)
                minb = max(minb, blue)

            if possible:
                res1 += game_id
            res2 += minr * ming * minb

        return Solution(res1, res2)


Solver.run()

File: nim/day02.nim
import re
import strutils

type Solution = tuple[part1: int, part2: int]

proc extract(s: string, re: Regex, default: int = 0): int =
  var matches: array[1, string]
  if s.find(re, matches) != -1: matches[0].parseInt else: default

proc solve(): Solution =
  var
    res1 = 0
    res2 = 0
    rTotal = 12
    gTotal = 13
    bTotal = 14

  for game in stdin.lines:
    var
      possible = true
      minR = 0
      minG = 0
      minB = 0
    
    let game_and_rest = game.split(re":")
    let game_id = game_and_rest[0].split(re" ")[1].parseInt

    for s in game_and_rest[1].split(";"):
      let
        red = s.extract(re"(\d+) red")
        green = s.extract(re"(\d+) green")
        blue = s.extract(re"(\d+) blue")

      minR = max(minR, red)
      minG = max(minG, green)
      minB = max(minB, blue)
      if red > rTotal or green > gTotal or blue > bTotal:
        possible = false


    if possible:
      res1 += game_id
    res2 += minR * minG * minB

  (res1, res2)

echo solve()
